base64ct/src/base64.rs

```diff
@@ -0,0 +1,229 @@
+//! # Base64
+//!
+//! Base64 encoding as described in [RFC 4648, section 4].
+//!
+//! [RFC 4648, section 4]: https://tools.ietf.org/html/rfc4648#section-4
+
+use crate::errors::{InvalidEncodingError, InvalidLengthError};
+use core::str;
+
+#[cfg(feature = "alloc")]
+use alloc::{string::String, vec::Vec};
+
+/// Encode
+pub fn encode(src: &[u8], dst: &mut [u8]) -> Result<usize, InvalidLengthError> {
+    if encoded_len(src) > dst.len() {
+        return Err(InvalidLengthError);
+    }
+
+    let mut src_offset: usize = 0;
+    let mut dst_offset: usize = 0;
+    let mut src_length: usize = src.len();
+
+    while src_length >= 3 {
+        encode_3bytes(
+            &src[src_offset..(src_offset + 3)],
```

Comment:
Why use the `while` loop instead of the approach used in the `b64` module? The latter is more idiomatic and allows compiler to remove bound checks. Same for `decode`.

Reply:
I tried to get this working but had a little bit of trouble. The logic is a bit different from the unpadded case.

Reply:
I managed to get it working with `chunks` instead of a loop, but I couldn't get it working with `chunks_exact`. Perhaps you could take another look.